src/Nri/Ui/Button/V9.elm

```diff
@@ -0,0 +1,887 @@
+large (ButtonOrLink config) =
+    ButtonOrLink { config | size = Large }
+
+
+
+-- BUTTON WIDTH
+
+
+type ButtonWidth
+    = WidthExact Int
+    | WidthUnbounded
+    | WidthFillContainer
+
+
+{-| Sizes for buttons and links that have button classes
+-}
+type ButtonSize
+    = Small
+    | Medium
+    | Large
+
+
+{-| Define a size in `px` for the button's total width.
+-}
+exactWidth : Int -> Attribute msg
+exactWidth inPx (ButtonOrLink config) =
+    ButtonOrLink { config | width = WidthExact inPx }
+
+
+{-| Leave the maxiumum width unbounded (there is a minimum width).
+-}
+unboundedWidth : Attribute msg
+unboundedWidth (ButtonOrLink config) =
+    ButtonOrLink { config | width = WidthUnbounded }
+
+
+{-| -}
+fillContainerWidth : Attribute msg
+fillContainerWidth (ButtonOrLink config) =
+    ButtonOrLink { config | width = WidthFillContainer }
+
+
+
+-- COLOR SCHEMES
+
+
+type ButtonStyle
+    = Primary
+    | Secondary
+    | Danger
+    | Premium
+      -- The following styles cannot be set directly.
+      -- Use a helper like `unfulfilled` to change the state instead.
+    | InactiveColors
+    | LoadingColors
+    | SuccessColors
+    | ErrorColors
+
+
+{-| -}
+primary : Attribute msg
+primary (ButtonOrLink config) =
+    ButtonOrLink { config | style = Primary }
+
+
+{-| -}
+secondary : Attribute msg
+secondary (ButtonOrLink config) =
+    ButtonOrLink { config | style = Secondary }
+
+
+{-| -}
+danger : Attribute msg
+danger (ButtonOrLink config) =
+    ButtonOrLink { config | style = Danger }
+
+
+{-| -}
+premium : Attribute msg
+premium (ButtonOrLink config) =
+    ButtonOrLink { config | style = Premium }
+
+
+
+-- BUTTON STATE
+
+
+type ButtonState
+    = Enabled
+    | Unfulfilled
+    | Disabled
+    | Error
+    | Loading
+    | Success
+
+
+{-| An enabled button. Takes the appearance of ButtonStyle.
+-}
+enabled : Attribute msg
+enabled (ButtonOrLink config) =
+    ButtonOrLink { config | state = Enabled }
+
+
+{-| A button which appears with the InactiveColors palette but is not disabled.
+-}
+unfulfilled : Attribute msg
+unfulfilled (ButtonOrLink config) =
+    ButtonOrLink { config | state = Unfulfilled }
+
+
+{-| A button which appears with the InactiveColors palette and is disabled.
+-}
+disabled : Attribute msg
+disabled (ButtonOrLink config) =
+    ButtonOrLink { config | state = Disabled }
+
+
+{-| A button which appears with the ErrorColors palette and is disabled.
+-}
+error : Attribute msg
+error (ButtonOrLink config) =
+    ButtonOrLink { config | state = Error }
+
+
+{-| A button which appears with the LoadingColors palette and is disabled.
+-}
+loading : Attribute msg
+loading (ButtonOrLink config) =
+    ButtonOrLink { config | state = Loading }
+
+
+{-| A button which appears with the SuccessColors palette and is disabled .
+-}
+success : Attribute msg
+success (ButtonOrLink config) =
+    ButtonOrLink { config | state = Success }
+
+
+
+-- INTERNALS
+
+
+{-| -}
+type alias Attribute msg =
```

Comment:
🎨 I think this should be a custom type instead of an alias, to keep this detail private.  Could be `type Attribute msg = Attribute (ButtonOrLink msg -> ButtonOrLink msg)` for now

Reply:
I agree with AV. I think this should be an opaque type.